blog/2020-03-01-CombiningBPMNWithOptaplanner.adoc

```diff
@@ -0,0 +1,611 @@
+when to terminate. Modify your `application.properties` to look like the following:
+
+[source,sh]
+----
+server.address=0.0.0.0
+
+spring.mvc.servlet.path=/docs
+
+resteasy.jaxrs.scan-packages=org.kie.**,com.company,http*
+
+# Solve for 5 seconds
+optaplanner.solver.termination.spent-limit=5s
+----
+
+== Creating the Constraints
+
+We can use the Constraint Stream API to create the constraints of our
+application in Java. We'll have two constraints:
+
+* No two passengers can be in the same seat.
+* Maximize the number of passengers who get seats they prefer.
+
+To do this, we create a `ConstraintProvider` that implement these constraints:
+
+[source,java]
+----
+public class FlightSeatingConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+            onlyOnePassengerPerSeat(constraintFactory),
+            seatTypePreferrence(constraintFactory)
+        };
+    }
+
+    private Constraint onlyOnePassengerPerSeat(ConstraintFactory constraintFactory) {
+        return constraintFactory.fromUniquePair(Passenger.class, Joiners.equal(p -> p.seat))
+                   .penalize("Only one passenger per seat", HardSoftScore.ONE_HARD);
+    }
+
+    private Constraint seatTypePreferrence(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Passenger.class)
+                   .join(Seat.class, Joiners.equal(p -> p.seat, s -> s))
+                   .filter((p,s) -> p.preferredSeatType != null && !p.preferredSeatType.equals(s.type))
+                   .penalize("Seat type preferrence", HardSoftScore.ONE_SOFT);
+    }
+
+}
+
+----
+
+== Testing the Flight Application
+
+Now we are ready to test out the Flight Application we just created!
+Run the following command in the terminal to start the server:
+
+```
+mvn clean package spring-boot:run
+```
+
+You can see the generated API methods by copying
+http://localhost:8080/docs/swagger.json to https://editor.swagger.io[The Swagger Editor].
+Let try creating a flight. Run the following command :
+
+```
+curl -X POST "localhost:8080/flightProcess" -H  "accept: application/json" -H  "Content-Type: application/json" -d "{\"flight\":{\"id\":\"string\",\"passengerList\":[],\"rowCount\":4,\"columnCount\":4}}"
+```
+
+```
+curl -X POST "localhost:8080/flightProcess" -H  "accept: application/json" -H  "Content-Type: application/json" -d "{\"flight\":{\"id\":\"string\",\"passengerList\":[],\"seatList\":[{\"seatIdentifier\":\"1A\",\"type\":\"WINDOW\"},{\"seatIdentifier\":\"1B\",\"type\":\"AISLE\"},{\"seatIdentifier\":\"1C\",\"type\":\"AISLE\"},{\"seatIdentifier\":\"1D\",\"type\":\"WINDOW\"},{\"seatIdentifier\":\"2A\",\"type\":\"WINDOW\"},{\"seatIdentifier\":\"2B\",\"type\":\"AISLE\"},{\"seatIdentifier\":\"2C\",\"type\":\"AISLE\"},{\"seatIdentifier\":\"2D\",\"type\":\"WINDOW\"}]}}"
+```
+
+You should get a response that looks like the following:
+[source]
```

Comment:
_Can_ be `[source,json]`. You'll get syntax highlight in IDEA, doesn't affect the rendered website.